app/controllers/admin_controller.rb

```diff
@@ -8,28 +8,118 @@ def index
     redirect_to(root_url) && return unless current_user.has_role? :admin
   end
 
+  def users
```

Comment:
Make sure to check for permissions before any other thing happens in this method, like: `redirect_to(root_url) && return unless current_user.has_role? :admin`

Reply:
Done.